### DISCUSSION TITLE
A Collection of Helpful Resources for Contributing to Projects

### DISCUSSION THREAD

**Discussion by anaximeno** [linuxmint/discussions#10] [Tutorials]
This document gathers helpful references for contributing to GTK projects, particularly within the Cinnamon/Linux Mint ecosystem.  This is a work in progress, so contributions and corrections are welcome!

**I. Core GTK & GJS Documentation:**

* **GTK API References:**
    * [GDK modules docs (C++)](https://api.gtkd.org/): Although in C++, these are helpful for understanding the underlying GDK modules used in other languages.
    * [GTK 3 Widgets Gallery](https://docs.gtk.org/gtk3/visual_index.html): Visual overview of common GTK 3 widgets (relevant for Linux Mint).
    * [GTK 4 Widgets Gallery](https://docs.gtk.org/gtk4/visual_index.html):  Reference for the newer GTK 4 widgets.
* **GJS (JavaScript for GNOME):**
    * [GJS Introduction](https://gjs.guide/guides/gjs/intro.html): Basic introduction to GJS, the foundation for **Cinnamon's JavaScript interpreter (CJS)**.
    * [GObject Usage in GJS](https://gitlab.gnome.org/GNOME/gjs/blob/master/doc/Mapping.md): Overview of using GObject with GJS.
    * [GJS Docs](https://gjs-docs.gnome.org/):  Documentation for GJS tools and APIs.

**II. Language-Specific Resources:**

* **Python:**
    * [Python GTK 3 Tutorial](https://python-gtk-3-tutorial.readthedocs.io/en/latest/): Introduction to GTK 3 application development in Python.
    * [PyGObject API Reference](https://lazka.github.io/pgi-docs/): Reference for PyGObject, the Python binding for GObject.
    * [Debian Python packaging guide](https://wiki.debian.org/Python/LibraryStyleGuide):  Style guide for packaging Python libraries for Debian (relevant for Linux Mint).
    * [Packaging Python Projects](https://packaging.python.org/en/latest/tutorials/packaging-projects/): Tutorial on packaging Python projects.
* **Vala:**
    * [Learn Vala and Gtk+ from Scratch (Video Playlist)](https://www.youtube.com/playlist?list=PLriKzYyLb28mn2lS3c5yqMHgLREi7kR9-):  Although focused on Vala, this provides practical examples of GTK usage.
    * [Vala Docummentation](https://valadoc.org/): Official documentation for Vala

**III. Linux Mint & Cinnamon Specific Resources:**

* [The Linux Mint Developer Guide](https://linuxmint-developer-guide.readthedocs.io/en/latest/): Essential information for contributing to Linux Mint.
* **Cinnamon Spice Repositories:**
    * [Applets](https://github.com/linuxmint/cinnamon-spices-applets)
    * [Extensions](https://github.com/linuxmint/cinnamon-spices-extensions)
    * [Desklets](https://github.com/linuxmint/cinnamon-spices-desklets)
    * [Actions](https://github.com/linuxmint/cinnamon-spices-actions)

**IV. Debian Packaging & Policies:**

* [Debian Developers' Manuals](https://www.debian.org/doc/devel-manuals): Debian policy and packaging information (relevant as Linux Mint is based on Debian/Ubuntu).

**V.  Useful Tools:**

* **Devhelp:** Browse installed documentation (e.g., `libgtk-3-doc`).
* **Meld:**  Compare files and directories for tracking changes.
* **D-feet:** Inspect and interact with D-Bus interfaces.
* **Glade:**  GUI designer for GTK applications.
 * **Git**: Essential for contributing to projects.
 * **Meson**: Build system used for some projects.

**VI. Code Examples:**

* [GJS Examples](https://github.com/optimisme/gjs-examples):  Example GTK applications written in GJS. Run examples on Linux Mint using `cjs nameOfTheExampleFile.js`.

**VII.  Graphics & Drawing:**

* [Cairo Documentation](https://www.cairographics.org/documentation):  Reference for Cairo, used for graphics and animations in Cinnamon.

**Contributions and Corrections Welcome!**  This document is a work in progress. Please suggest additions or corrections with explanations.

**Comment by JosephMcc** [comment]
You left out the devhelp application. When you install doc packages such as libgtk-3-doc it's documentation will show up in the application.

**Reply by clefebvre** [reply]
Devhelp, meld (visual diff tool), d-feet (for dbus), glade (for .ui files), sublime-text, sublime-merge...

**Comment by hsbasu** [comment]
Debian Developers' Manuals - https://www.debian.org/doc/devel-manuals
Debian Policy Manual - https://www.debian.org/doc/debian-policy/
Debian New Maintainers' Guide - https://www.debian.org/doc/manuals/maint-guide/index.en.html
debian Python packaging guide - https://wiki.debian.org/Python/LibraryStyleGuide

Packaging Python Projects/codes - https://packaging.python.org/en/latest/tutorials/packaging-projects/
